### DISCUSSION TITLE
Feature Request: TLS support

### DISCUSSION THREAD

**Discussion by limpens** [jziolkowski/tdm#9] [Ideas]
Please add the possibility to use a TLS enabled broker.

**Comment by jziolkowski** [comment]
I'll look into it, thanks for the suggestion.

**Comment by ScrewLooseDan** [comment]
I agree this would be nice to have.

**Comment by jziolkowski** [comment]
I've added a test branch: tls

Please pull and test if TLS connection is working as expected.

**Comment by limpens** [comment]
Thank you for the changes.

- Looks like the selected certificate is not saved in the ~/TDM/tdm.cfg
- when selecting a ca-certificate as certificate (like: mosquitto_sub -h _mqtt-server-fqdn_ -p 8883 -u _mqtt-user_ -P _mqtt-pass_ -t \#  -v  --cafile **/etc/mosquitto/ca_certificates/ca.crt**) the broker reports a SSL error: OpenSSL Error: error:1408F10B:SSL routines:ssl3_get_record:wrong version number

Eric.

**Comment by jziolkowski** [comment]
Thanks for the input. The first error I'll look into it in a few mins.
I take it your certificate works with other clients?

**Comment by limpens** [comment]
Yes, that certificate works with a commandline mosquitto_sub
like: mosquitto_sub -h mqtt.domain.tld -p 8883 -u mqtt-user -P mqtt-pass -t
\#  -v  --cafile /etc/mosquitto/ca_certificates/ca.crt

**Comment by jziolkowski** [comment]
Ok. I've fixed cert path saving and pushed extra parameters for TSL. Please try.

**Comment by limpens** [comment]
Saving (and loading) of the selected certificated appears to be working
fine.

The TLS connection is not yet completing successfully though. Tried to
change some of the parameters in self.mqtt.m_client.tls_set(), but no
success on my end.

**Comment by jziolkowski** [comment]
Same error?

**Comment by limpens** [comment]
Yes, this is the exact message in the mosquitto.log file:

1551171767: New connection from 192.168.10.254 on port 8883.
1551171767: OpenSSL Error: error:1408F10B:SSL
routines:ssl3_get_record:wrong version number
1551171767: Socket error on client <unknown>, disconnecting.

The same would happen when using mosquitto_sub without a ca certificate
file.

**Comment by jziolkowski** [comment]
Ok, I'm getting more testers on board to check if it's an isolated issue or something with the implementation.

**Comment by jziolkowski** [comment]
Also, which python version are you using?

**Comment by limpens** [comment]
Python 3.6.7 and python3-paho-mqtt 1.3.1-1 (Ubuntu 18.04.2)

**Comment by jziolkowski** [comment]
I've tried with my own server and the issue is quite complex. Related to openssl libs on the server, on the client, and these python ssl module was built against. I must put it on hold for now.

**Comment by Thomseeen** [comment]
Repository seems quite quiet, there is no ssl-branch anymore? Just came across this and might look into it myself if there is no solution yet.

**Comment by alexpilotti** [comment]
I'm using TLS and got it to work by just adding:

`self.m_client.tls_set(ca_certs="/path/to/root.pem")`

before:

https://github.com/jziolkowski/tdm/blob/9b3edf58da3cbdf6435ead30e4a9d0f379073bdc/Util/mqtt.py#L119
